routes/user.js

```diff
@@ -41,312 +41,280 @@ var config = require('config')
-                user.resetPasswordToken = undefined;
-                user.resetPasswordExpires = undefined;
-
-                user.save(function(err) {
-                    req.logIn(user, function(err) {
-                        done(err, user);
+router.route('/reset/:token')
+    .all(passportFunction.notLoggedIn)
+    .get(function(req, res, next) {
+        User.findOne({
+            resetPasswordToken: req.params.token,
+            resetPasswordExpires: {
+                $gt: Date.now()
+            }
+        }, function(err, user) {
+            if (!user) {
+                req.flash('error', 'Password reset token is invalid or has expired.');
+                return res.redirect('/user/forgot');
+            }
+            res.render('reset', {
+                user: req.user
+            });
+        });
+    })
+    .post(function(req, res, next) {
+        async.waterfall([
+            function(done) {
+                User.findOne({
+                    resetPasswordToken: req.params.token,
+                    resetPasswordExpires: {
+                        $gt: Date.now()
+                    }
+                }, function(err, user) {
+                    if (!user) {
+                        req.flash('error', 'Password reset token is invalid or has expired.');
+                        return res.redirect('back');
+                    }
+
+                    user.password = user.generateHash(req.body.password);
+                    user.resetPasswordToken = undefined;
+                    user.resetPasswordExpires = undefined;
+
+                    user.save(function(err) {
+                        req.logIn(user, function(err) {
+                            done(err, user);
+                        });
                     });
                 });
-            });
-        },
-        function(user, done) {
-            var smtpTransport = nodemailer.createTransport('SMTP', {
-                service: 'SendGrid',
-                auth: {
-                    user: process.env.SENDGRID_USER || config.get('emailVerification.user'),
-                    pass: process.env.SENDGRID_PASS || config.get('emailVerification.pass')
-                }
-            });
-            var mailOptions = {
-                to: user.email,
-                from: 'noreply@theotech.org',
-                subject: 'Your password has been changed',
-                text: 'Hello,\n\n' +
-                    'This is a confirmation that the password for your account ' + user.email + ' has just been changed.\n'
-            };
-            smtpTransport.sendMail(mailOptions, function(err) {
-                req.flash('success', 'Success! Your password has been changed.');
-                done(err);
-            });
-        }
-    ], function(err) {
-        res.redirect('/');
+            },
+            function(user, done) {
+                var smtpTransport = nodemailer.createTransport('SMTP', {
```

Comment:
Also consider, wrapping up this smtpTransport into an object that you can use simply. 

Something like mailer.send(mailOptions, callback);
It'll make your code cleaner/more modular.